src/Run.purs

```diff
@@ -104,42 +105,80 @@ send
   → Run r a
 send = RunM <<< liftF
 
+-- | Casts some set of effects to a wider set of effects via a left-biased
+-- | union.
+expand
+  ∷ ∀ r1 r2 rx a
+  . Union r1 rx r2
+  ⇒ Run r1 a
+  → Run r2 a
+expand = unsafeCoerce
+
 -- | Extracts the value from a fully interpreted program.
 run ∷ ∀ a. Run () a → a
 run = unwrap >>> runFree \_ → unsafeCrashWith "Control.Monad.Run: the impossible happened"
 
--- | Extracts the value from a program with only base effects.
+unBase ∷ ∀ f. VariantF (base ∷ FProxy f) ~> f
+unBase = case_ # on (SProxy ∷ SProxy "base") id
+
+-- | Extracts the value from a program with only base effects. This assumes
+-- | stack safety under Monadic recursion.
 runBase
   ∷ ∀ f a
+  . Monad f
+  ⇒ Run (base ∷ FProxy f) a
+  → f a
+runBase = unwrap >>> loop
+  where
+  loop = resume' (\b f → f <$> unBase b >>= loop) pure
+
+-- | Extracts the value from a program with only base effects using `MonadRec`
+-- | to preserve stack safety.
+runBaseRec
+  ∷ ∀ f a
   . MonadRec f
   ⇒ Run (base ∷ FProxy f) a
   → f a
-runBase = unwrap >>> foldFree go
+runBaseRec = unwrap >>> foldFree (case_ # on (SProxy ∷ SProxy "base") id)
+
+-- | Interprets the base effect into some Monad `m` via continuation passing.
+runBaseWithCont
+  ∷ ∀ f m a b
+  . Monad m
+  ⇒ (f (Unit → m b) → m b)
+  → (a → m b)
+  → Run (base ∷ FProxy f) a
+  → m b
+runBaseWithCont k1 k2 = unwrap >>> loop
   where
-  go ∷ VariantF (base ∷ FProxy f) ~> f
-  go = case_ # on (SProxy ∷ SProxy "base") id
+  loop = resume' (\b f → k1 (unBase $ (\b' _ → loop (f b')) <$> b)) k2
 
 -- | Interprets an effect functor into the base effect, eliminating the label.
-interpret
-  ∷ ∀ sym f m r1 r2 r3 a
-  . RowCons sym (FProxy f) r2 r1
-  ⇒ RowCons "base" (FProxy m) r2 r3
+runWithBase
+  ∷ ∀ sym f m r1 r2 a
+  . RowCons sym (FProxy f) (base ∷ FProxy m | r2) r1
```

Comment:
This now means that the `Run r a` you start with has to already have a `base` row (or be unifiable with one that does), right? Is that intentional? Seems like it would force all of the rows to be open if you ever wanted to use `runWithBase`.

Reply:
Yes, this is intentional, and that's because the old way is unsafe. Previously it would remove the old label and _add_ a base label. This means if you did it multiple times, you could have multiple base effects existing at runtime which would be unsafe. That's also why I added `expand`, which lets you widen the set of rows via a `Union` constraint. This means you could take a program with effects A and B, and widen it to include C, because we know that it _doesn't_ contain C.

It may be possible to do something fancy with `RowToList` and instance chains where we do an effect label upsert-like thing.

Reply:
Gotcha!